internal/v5/resource.go

```diff
@@ -0,0 +1,176 @@
+	}
+	resource.Sort(resources)
+	return resources, nil
+}
+
+// POST id/resource/name
+// https://github.com/juju/charmstore/blob/v5/docs/API.md#post-idresourcesname
+//
+// GET  id/resource/name[/revision]
+// https://github.com/juju/charmstore/blob/v5/docs/API.md#get-idresourcesnamerevision
+func (h *ReqHandler) serveResources(id *router.ResolvedURL, w http.ResponseWriter, req *http.Request) error {
+	// Resources are "published" using "POST id/publish" so we don't
+	// support PUT here.
+	// TODO(ericsnow) Support DELETE to remove a resource (like serveArchive)?
+	switch req.Method {
+	case "GET":
+		return h.serveDownloadResource(id, w, req)
+	case "POST":
+		return h.serveUploadResource(id, w, req)
+	default:
+		return errgo.WithCausef(nil, params.ErrMethodNotAllowed, "%s not allowed", req.Method)
+	}
+}
+
+func (h *ReqHandler) serveDownloadResource(id *router.ResolvedURL, w http.ResponseWriter, req *http.Request) error {
+	rid, err := parseResourceId(strings.TrimPrefix(req.URL.Path, "/"))
+	if err != nil {
+		return errgo.WithCausef(err, params.ErrNotFound, "")
+	}
+	ch, err := h.entityChannel(id)
+	if err != nil {
+		return errgo.Mask(err, errgo.Is(params.ErrNotFound))
+	}
+	r, err := h.Store.ResolveResource(id, rid.Name, rid.Revision, ch)
+	if err != nil {
+		return errgo.Mask(err, errgo.Is(params.ErrNotFound))
+	}
+	blob, err := h.Store.OpenResourceBlob(r)
+	if err != nil {
+		return errgo.Notef(err, "cannot open resource blob")
+	}
+	defer blob.Close()
+	header := w.Header()
+	setArchiveCacheControl(w.Header(), h.isPublic(id))
+	header.Set(params.ContentHashHeader, blob.Hash)
+
+	// TODO(rog) should we set connection=close here?
+	// See https://codereview.appspot.com/5958045
+	serveContent(w, req, blob.Size, blob)
+	return nil
+}
+
+func (h *ReqHandler) serveUploadResource(id *router.ResolvedURL, w http.ResponseWriter, req *http.Request) error {
+	if id.URL.Series == "bundle" {
+		return errgo.WithCausef(nil, params.ErrForbidden, "cannot upload a resource to a bundle")
+	}
+	name := strings.TrimPrefix(req.URL.Path, "/")
+	if !validResourceName(name) {
+		return badRequestf(nil, "invalid resource name")
+	}
+	hash := req.Form.Get("hash")
+	if hash == "" {
+		return badRequestf(nil, "hash parameter not specified")
+	}
+	if req.ContentLength == -1 {
+		return badRequestf(nil, "Content-Length not specified")
+	}
+	e, err := h.Cache.Entity(&id.URL, charmstore.FieldSelector("charmmeta"))
+	if err != nil {
+		// Should never happen, as the entity will have been cached
+		// when the charm URL was resolved.
+		return errgo.Mask(err, errgo.Is(params.ErrNotFound))
+	}
+	r, ok := e.CharmMeta.Resources[name]
+	if !ok {
+		return errgo.WithCausef(nil, params.ErrForbidden, "resource %q not found in charm metadata", name)
+	}
+	if r.Type != resource.TypeFile {
+		return errgo.WithCausef(nil, params.ErrForbidden, "non-file resource types not supported")
+	}
+	if filename := req.Form.Get("filename"); filename != "" && path.Ext(filename) != "" {
```

Comment:
I think that it's reasonable that if you have a filename in the charm metadata with
no extension that it's extension-agnostic and can be used with any file.
Otherwise there's no way of having a resource that's just a blob of
format-independent data.

Similarly for uploading a file with no extension, I think, although that's
perhaps less important. Tbh I don't think this file extension checking
thing is that important or useful. It won't even let me upload a .tgz to a .tar.gz
file.